### ISSUE TITLE
What part of transform should be updated when interact with layers in `mode == "transform"`

### ISSUE THREAD

**Issue by Czaki** [napari/napari#6446]
### Overview of design need

Currently, modifying a Layer that is in transform mode by mouse interaction modifies the `layer.affine`. It means that `layer.scale`, `layer.rotate` become untouched. As affine operates in `physical2world` space where all properties like transform, scale operates in `data2physical` 

It is non-intuitive for me so I introduce bug in #6438. I think that it may also be non intuitive for plugin creators who may want to use the mouse to align data and then read transform, scale etc from the layer. That will fail. 

From my point we should do one of three things:

1) split `transform` mode on `transform_data`, `transform_affine` to allow the user decide what he wants to update
2) change `transform` to operate on `data2physical` 
3) clarify the situation in the documentation and explain that the user needs to use `layer.transform + layer.affine.transform`, `np.dot(layer.rotate, layer.affine.rotate)`. etc. 

cc @brisvag @psobolewskiPhD @jni @haesleinhuepf

**Comment by brisvag** [comment]
I'll report what my recollection of the last discussions about this is:

I think the idea was that `data2physical` is something that comes from the data itself (pixel size, beam shift in multi-shot microscopy, etc), and it represents how to go from "data space" to a space with physical meaning. The affine afterwards should be for extra modifications on top, like registration to other layers, or adjustments from the user like dragging/rotating the layer.

If you ask me, I think we should have only one affine and not this combination, but I remember some people advocating for this separation.

If we keep the two, I think the best option is 3... 1 just adds to the confusion, and 2 will catch the other half of the people offguard.

**Comment by jo-mueller** [comment]
Hi @Czaki @brisvag , took me a bit to find this issue but I think it would be really neat to have this fixed. It was kind of confusing to me when I was working with some shapes data layers lately and found that I could move them around freely (cool!) without apparently changing the translate/rotate properties.

As for the options listed above, I guess changing the actual data in the layer could get tricky (especially for images a lot of interpolations would have to be done), so I'd opt for updating the `layer.transform`, `layer.scale` and `layer.rotate` properties.

It's good to know that `affine` is the property I should have looked for - I figure it wouldn't be that hard to split `layer.affine` into a translate/scale/rotate component and simply overwrite the respective properties?

*Edit*: Just realized myself that this would lead to stacked application of translate/scale/rotate properties during the usage of the interactive edit. Tricky :/

**Comment by brisvag** [comment]
Yeah it's not obvios how to solve this so there are no unintuitive scenarios. Either we do a *much* better job at communicating what transformations each layer goes through and whhat they mean, or we should unify everything under the same affine. However, I know that I got pushback on this before, as folks wanted the separation between transforms that come from the data *collection* versus transforms that are applied in *processing*.

**Comment by psobolewskiPhD** [comment]
I totally agree that the way it is right now is super confusing.
Ran into this with https://github.com/napari/napari/pull/8098
where layer.affine layer.scale and layer.translate are sometimes not and sometimes partially synced.
At a minimum this should be much better documented!

**Comment by Czaki** [comment]
The affine is a combination of scale, translate, and rotate saved as $n+1 \times n+1$ matrix. 

We may introduce the layer method that will update `scale`, `transform` and `rotate` based on affine content and change `affine` to "identity".

**Comment by TimMonko** [comment]
> However, I know that I got pushback on this before, as folks wanted the separation between transforms that come from the data _collection_ versus transforms that are applied in _processing_.

What does this mean, in practice though? Especially if, as Peter said, they are sometimes in sync.

I think it would make sense for setting one of these sync and updates all the rest. As Grzegorz mentioned, it can be all be passed into and update the affine, and get the results back from that. So it seems possible to sync?

**Comment by brisvag** [comment]
> What does this mean, in practice though? Especially if, as Peter said, they are sometimes in sync.

I think @psobolewskiPhD got confused: they are *not* related, ever.

If you look at the code from the base `Layer`, this is for example `scale`:

https://github.com/napari/napari/blob/bfb0ab5308d3ba08090d12f5e778949de3560064/src/napari/layers/base/base.py#L839-L842

The same is true for rotate, shear and translate.

These are all from the `data2physical` affine matrix, while `Layer.affine` refers to the `physical2world` affine.

https://github.com/napari/napari/blob/bfb0ab5308d3ba08090d12f5e778949de3560064/src/napari/layers/base/base.py#L901-L904

They are all instantiated here:

https://github.com/napari/napari/blob/bfb0ab5308d3ba08090d12f5e778949de3560064/src/napari/layers/base/base.py#L407-L425

---

If these are ever "synced", it's either by chance or by mistake, or a bug.

> The affine is a combination of scale, translate, and rotate saved as n + 1 × n + 1 matrix.

This is *an affine*, but not *the* `Layer.affine`.

**Comment by jni** [comment]
@brisvag has all the right history here. The separation is very important and should not go away, but should be signposted better both in code and in docs. For more on the *future* of transforms see #3783, and OME NGFF's RFC-5: https://ngff.openmicroscopy.org/rfc/5/index.html

**Comment by jo-mueller** [comment]
Maybe as a starting point for the documentation - what would be the correct way to retrieve the position of an object in world coordinates given two sets of transformation parameters (i.e., from data and from processing)? For most properties I guess it's relatively clear as @Czaki wrote above:

```python

translate_net = layer.translate + layer.affine.translate
rotate_net = layer.rotate + layer.affine.rotate
scale_net = layer.scale * layer.affine.scale
shear_net = layer.shear + layer.affine.shear

position_in_data_coordinates = some_position

# aggregate components into affine matrix
# is there an easy way to do this? I.e., to create the `Affine` matrix class from napari?

...

# get world coordinates
world_coordinate_position = affine_matrix_net @ position_in_data_coordinates 
```

**Comment by brisvag** [comment]
> is there an easy way to do this? I.e., to create the `Affine` matrix class from napari?

Yes:

```py
from napari.utils.transforms import Affine
```

```py
Affine(
    scale=(1.0, 1.0),
    translate=(0.0, 0.0),
    *,
    affine_matrix=None,
    axis_labels: collections.abc.Sequence[str] | None = None,
    linear_matrix=None,
    name=None,
    ndim=None,
    rotate=None,
    shear=None,
    units: collections.abc.Sequence[str | pint.registry.Unit] | None = None,
) -> None
```

Which is easily used by calling:

```py
my_affine(position_in_data_coordinates)
```

**Comment by jo-mueller** [comment]
@zoccoler I think this at least solves what we were wrestling with the other day!

**Comment by zoccoler** [comment]
Hi @jo-mueller ,

I have this fixed in [my fork](https://github.com/zoccoler/napari-clusters-plotter/blob/c982779aa8ff7521636ce6f2283a480cba687550/src/napari_clusters_plotter/_new_plotter_widget.py#L852). I basically get the affine matrix from the `Affine` class via the `.affine_matrix` attribute, but I built the other affine matrix from the parameters based on their definitions. I will try to replace that with the code above, as it simplifies code and avoid duplicating it, and check if it still works :)

**Comment by zoccoler** [comment]
It works! That's how I am doing it so far:

```python
rotate = layer.rotate
scale = layer.scale
shear = layer.shear
translate = layer.translate
affine_data2physical = Affine(
    rotate=rotate,
    scale=scale,
    shear=shear,
    translate=translate,
).affine_matrix
affine_physical2world = layer.affine.affine_matrix
# Combine the two affine transformations
affine_net = affine_data2physical @ affine_physical2world
```
If I apply `affine_net` to a data point and move the camera there, it preserves point of view while considering the data physical space.
